File: bench/bench_ocaml/lib/float.ml
open Common

(* #import "math-lib-core/math.mligo" "Math" *)

(* n = a * 10^b *)
type t = { val_ : int; pow : int }

let new' (val_ : int) (pow : int) : t = { val_; pow }

let inverse (a : t) : t =
  {
    val_ = Z.one * Math.power (ten, eighteen) / a.val_;
    pow = (a.pow * minus_one) - eighteen;
  }

let add (a : t) (b : t) : t =
  if a.pow < b.pow then
    {
      val_ = (b.val_ * Math.power (ten, abs (b.pow - a.pow))) + a.val_;
      pow = a.pow;
    }
  else
    {
      val_ = (a.val_ * Math.power (ten, abs (a.pow - b.pow))) + b.val_;
      pow = b.pow;
    }

let sub (a : t) (b : t) : t =
  if a.pow < b.pow then
    {
      val_ = a.val_ - (b.val_ * Math.power (ten, abs (b.pow - a.pow)));
      pow = a.pow;
    }
  else
    {
      val_ = (a.val_ * Math.power (ten, abs (a.pow - b.pow))) - b.val_;
      pow = b.pow;
    }

let lt (a : t) (b : t) : bool =
  if a.val_ < Z.zero && b.val_ > Z.zero then true
  else if a.val_ > Z.zero && b.val_ < Z.zero then false
  else
    let diff = sub a b in
    diff.val_ < Z.zero

let lte (a : t) (b : t) : bool =
  if a.val_ < Z.zero && b.val_ > Z.zero then true
  else if a.val_ > Z.zero && b.val_ < Z.zero then false
  else
    let diff = sub a b in
    diff.val_ <= Z.zero

let gte (a : t) (b : t) : bool =
  if a.val_ >= Z.zero && b.val_ < Z.zero then true
  else if a.val_ <= Z.zero && b.val_ > Z.zero then false
  else
    let diff = sub a b in
    diff.val_ >= Z.zero

let gt (a : t) (b : t) : bool =
  if a.val_ > Z.zero && b.val_ < Z.zero then true
  else if a.val_ < Z.zero && b.val_ > Z.zero then false
  else
    let diff = sub a b in
    diff.val_ > Z.zero

let mul (a : t) (b : t) : t = { val_ = a.val_ * b.val_; pow = a.pow + b.pow }

let div (a : t) (b : t) : t =
  {
    val_ = a.val_ * Math.power (ten, eighteen) / b.val_;
    pow = a.pow - b.pow - eighteen;
  }

let modulo (a : t) (b : t) : t =
  let rec compute ((a, b) : t * t) : t =
    if lt a b then a else compute (sub a b, b)
  in
  compute (a, b)

let resolve (a : t) (prec : nat) : int =
  let resolve_positif (a : t) (prec : nat) : int =
    if a.pow > Z.zero then
      a.val_
      * Math.power (ten, abs a.pow)
      * Math.power (ten, prec)
    else
      a.val_
      * Math.power (ten, prec)
      / Math.power (ten, abs (minus_one * a.pow))
  in
  if a.val_ < Z.zero then minus_one * resolve_positif (new' (int' (abs a.val_)) a.pow) prec
  else resolve_positif a prec

File: bench/bench_kotlin/src/main/kotlin/math/Float.kt
package math

import math.lib.Math
import math.lib.Math.EIGHTEEN
import java.math.BigInteger
import java.math.BigInteger.ONE
import java.math.BigInteger.TEN
import java.math.BigInteger.ZERO

interface Float : Repr<Float.T> {

    data class T(val value: BigInteger, val pow: BigInteger) {
        override fun toString(): String = "$value*10^$pow"
    }


    fun new(v: Long): T = T(BigInteger.valueOf(v), ZERO)
    fun new(v: Long, p: Long): T = T(BigInteger.valueOf(v), BigInteger.valueOf(p))

    override fun abs(a: T): T = T(a.value.abs(), a.pow)

    override fun inverse(a: T): T =
        T(Math.power(TEN, EIGHTEEN) / a.value, (a.pow * -ONE) - EIGHTEEN)

    override fun add(a: T): (T) -> T = { b ->
        if (a.pow < b.pow) {
            T(b.value * Math.power(TEN, (b.pow - a.pow).abs()) + a.value, a.pow)
        } else {
            T(a.value * Math.power(TEN, (a.pow - b.pow).abs()) + b.value, b.pow)
        }
    }

    override fun sub(a: T): (T) -> T = { b ->
        if (a.pow < b.pow) {
            T(a.value - b.value * Math.power(TEN, (b.pow - a.pow).abs()), a.pow)
        } else {
            T(a.value * Math.power(TEN, (a.pow - b.pow).abs()) - b.value, b.pow)
        }
    }

    override fun lt(a: T): (T) -> Boolean = { b ->
        if (a.value < ZERO && b.value > ZERO) {
            true
        } else if (a.value > ZERO && b.value < ZERO) {
            false
        } else {
            sub(a)(b).value < ZERO
        }
    }

    override fun lte(a: T): (T) -> Boolean = { b ->
        if (a.value < ZERO && b.value > ZERO) {
            true
        } else if (a.value > ZERO && b.value < ZERO) {
            false
        } else {
            sub(a)(b).value <= ZERO
        }
    }

    override fun mul(a: T): (T) -> T = { b ->
        T(a.value * b.value, a.pow + b.pow)
    }

    override fun div(a: T): (T) -> T = { b ->
        T(a.value * Math.power(TEN, EIGHTEEN) / b.value, a.pow - b.pow - EIGHTEEN)
    }


    override fun resolve(a: T): (BigInteger) -> BigInteger = { prec ->
        fun resolve_positive(a: T, prec: BigInteger): BigInteger =
            if (a.pow > ZERO) {
                a.value * Math.power(TEN, a.pow.abs()) * Math.power(TEN, prec)
            } else {
                a.value * Math.power(TEN, prec) / Math.power(TEN, (-ONE * a.pow).abs())
            }

        if (a.value < ZERO) {
            -ONE * resolve_positive(T(a.value.abs(), a.pow), prec)
        } else {
            resolve_positive(a, prec)
        }
    }

}